app/frontend/src/features/messages/groupchats/GroupChatView.test.tsx

```diff
@@ -0,0 +1,172 @@
+import { render, screen, within } from "@testing-library/react";
+import { Empty } from "google-protobuf/google/protobuf/empty_pb";
+import React from "react";
+import {
+  mockAllIsIntersecting,
+  mockIsIntersecting,
+} from "react-intersection-observer/test-utils";
+import { Route } from "react-router-dom";
+import { messagesRoute } from "../../../AppRoutes";
+import { service } from "../../../service";
+import messageData from "../../../test/fixtures/messages.json";
+import { getHookWrapperWithClient } from "../../../test/hookWrapper";
+import {
+  getGroupChatMessages,
+  getUser,
+} from "../../../test/serviceMockDefaults";
+import { addDefaultUser, MockedService, wait } from "../../../test/utils";
+import { MARK_LAST_SEEN_TIMEOUT } from "../constants";
+import GroupChatView from "./GroupChatView";
+
+const getGroupChatMock = service.conversations.getGroupChat as MockedService<
+  typeof service.conversations.getGroupChat
+>;
+const getUserMock = service.user.getUser as jest.Mock;
+
+const getGroupChatMessagesMock = service.conversations
+  .getGroupChatMessages as MockedService<
+  typeof service.conversations.getGroupChatMessages
+>;
+const baseGroupChatMockResponse = {
+  groupChatId: 1,
+  title: "Test group chat",
+  memberUserIdsList: [1, 2, 3],
+  adminUserIdsList: [1],
+  onlyAdminsInvite: false,
+  isDm: false,
+  unseenMessageCount: 4,
+  lastSeenMessageId: 1,
+  latestMessage: {
+    messageId: 5,
+    authorUserId: 3,
+    text: { text: "In 2 hours?" },
+    time: { seconds: 1577900000, nanos: 0 },
+  },
+};
+
+const markLastSeenGroupChatMock = service.conversations
+  .markLastSeenGroupChat as MockedService<
+  typeof service.conversations.markLastSeenGroupChat
+>;
+
+// TODO: tests involving these mutations - maybe these can be localised only
```

Comment:
I thought I'd focus on the mark message as seen tests (which would have been quite hard without `react-intersection-observer`, the legend who wrote the mocks for `IntersectionObserver` as well that'd have taken me too long otherwise haha) for this. Happy to return to this later in a separate PR considering this area is still being tweaked?

Reply:
What are "these mutations?"

Reply:
As in tests that will use the mutations below this comment. So try sending a message through the UI, leave the group chat etc (and a few more I probably missed since your recent PR to add in the other missing actions)